File: go/pl/loom/internal/ui/markdown.go
// Copyright (c) 2026 The Authors. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Authors: liubang (it.liubang@gmail.com)
// Created: 2026/07/24

package ui

import (
	"crypto/sha256"
	"encoding/hex"
	"strings"
	"sync"

	"github.com/charmbracelet/glamour"
	"github.com/charmbracelet/glamour/styles"
)

// markdownCacheCap bounds the render cache. When exceeded the cache resets
// wholesale: chat transcripts are append-mostly, so a cold rebuild stays
// cheap while the map never grows without limit.
const markdownCacheCap = 512

// markdownRenderer wraps glamour with loom's theme awareness and a
// per-(content, width, style) render cache. Finalized assistant blocks are
// immutable, so caching is safe and avoids re-rendering the full transcript
// on every frame.
type markdownRenderer struct {
	mu        sync.Mutex
	renderers map[markdownRendererKey]*glamour.TermRenderer
	cache     map[markdownCacheKey]string
}

type markdownRendererKey struct {
	wordWrap int
	profile  string // "dark" | "light" | "notty"
}

type markdownCacheKey struct {
	digest   string
	wordWrap int
	profile  string
}

func newMarkdownRenderer() *markdownRenderer {
	return &markdownRenderer{
		renderers: make(map[markdownRendererKey]*glamour.TermRenderer),
		cache:     make(map[markdownCacheKey]string),
	}
}

// render converts markdown into a styled terminal string. wordWrap <= 0
// falls back to glamour's default width (used when the terminal size is not
// known yet); an empty profile selects the Everforest dark style.
func (r *markdownRenderer) render(content string, wordWrap int, profile string) string {
	if wordWrap < 0 {
		wordWrap = 0
	}
	if profile == "" {
		profile = "dark"
	}
	content = escapeLoneTildes(content)
	sum := sha256.Sum256([]byte(content))
	key := markdownCacheKey{digest: hex.EncodeToString(sum[:]), wordWrap: wordWrap, profile: profile}

	r.mu.Lock()
	defer r.mu.Unlock()
	if rendered, ok := r.cache[key]; ok {
		return rendered
	}
	renderer, err := r.termRenderer(markdownRendererKey{wordWrap: wordWrap, profile: profile})
	if err != nil {
		return content
	}
	rendered, err := renderer.Render(content)
	if err != nil {
		return content
	}
	// Strip the document-level blank lines without eating meaningful leading
	// indentation of the first content line (indented code, nested lists).
	rendered = strings.TrimLeft(rendered, "\n")
	rendered = strings.TrimRight(rendered, " \n\t")
	if len(r.cache) >= markdownCacheCap {
		r.cache = make(map[markdownCacheKey]string)
	}
	r.cache[key] = rendered
	return rendered
}

func (r *markdownRenderer) termRenderer(key markdownRendererKey) (*glamour.TermRenderer, error) {
	if renderer, ok := r.renderers[key]; ok {
		return renderer, nil
	}
	var options []glamour.TermRendererOption
	if key.wordWrap > 0 {
		options = append(options, glamour.WithWordWrap(key.wordWrap))
	}
	switch key.profile {
	case "notty":
		// notty is glamour's ASCII style: no ANSI sequences at all, which is
		// the only safe choice for NO_COLOR and dumb terminals.
		options = append(options, glamour.WithStandardStyle(styles.NoTTYStyle))
	case "light":
		options = append(options, glamour.WithStyles(styles.LightStyleConfig))
	default:
		options = append(options, glamour.WithStylesFromJSONBytes([]byte(everforestStyleJSON)))
	}
	renderer, err := glamour.NewTermRenderer(options...)
	if err != nil {
		return nil, err
	}
	r.renderers[key] = renderer
	return renderer, nil
}

// escapeLoneTildes escapes every single "~" as "\~" so goldmark's GFM
// strikethrough (which accepts both ~ and ~~) cannot strike through
// temperature ranges and similar everyday text ("31~33°C，体感 38~41°C"
// would otherwise render crossed out). "~~" pairs pass through
// untouched, keeping deliberate strikethrough available. Code spans and
// fenced code blocks are exempt: an escape backslash would render
// literally there. Indented code blocks are not detected (rare in
// model output; fences are the norm).
func escapeLoneTildes(s string) string {
	if !strings.Contains(s, "~") {
		return s
	}
	var b strings.Builder
	b.Grow(len(s) + 8)
	inFence := false
	for line := range strings.Lines(s) {
		trimmed := strings.TrimLeft(line, " \t")
		if strings.HasPrefix(trimmed, "```") || strings.HasPrefix(trimmed, "~~~") {
			inFence = !inFence
			b.WriteString(line)
			continue
		}
		if inFence {
			b.WriteString(line)
			continue
		}
		escapeLoneTildesInLine(&b, line)
	}
	return b.String()
}

// escapeLoneTildesInLine escapes lone tildes in one line, leaving inline
// code spans (backtick runs) untouched.
func escapeLoneTildesInLine(b *strings.Builder, line string) {
	i := 0
	for i < len(line) {
		if line[i] == '`' {
			j := i
			for j < len(line) && line[j] == '`' {
				j++
			}
			end := findClosingTicks(line, j, j-i)
			if end < 0 {
				// No closing run: the rest of the line is literal.
				b.WriteString(line[i:])
				return
			}
			b.WriteString(line[i:end])
			i = end
			continue
		}
		if line[i] == '~' {
			if i+1 < len(line) && line[i+1] == '~' {
				b.WriteString("~~")
				i += 2
				continue
			}
			b.WriteString("&#126;")
			i++
			continue
		}
		b.WriteByte(line[i])
		i++
	}
}

// findClosingTicks locates the closing run of n backticks starting the
// search at from, returning the index just past it (-1 when absent).
func findClosingTicks(s string, from, n int) int {
	idx := strings.Index(s[from:], strings.Repeat("`", n))
	if idx < 0 {
		return -1
	}
	return from + idx + n
}

// everforestStyleJSON is the loom dark theme as a glamour style sheet, using
// the same Everforest Dark Hard palette as the rest of the TUI.
const everforestStyleJSON = `{
  "document": {
    "block_prefix": "\n",
    "block_suffix": "\n",
    "color": "#d3c6aa",
    "margin": 0
  },
  "block_quote": {
    "indent": 1,
    "indent_token": "│ ",
    "color": "#859289",
    "italic": true
  },
  "paragraph": {},
  "list": {
    "level_indent": 2
  },
  "heading": {
    "block_suffix": "\n",
    "color": "#7fbbb3",
    "bold": true
  },
  "h1": {
    "prefix": "# ",
    "color": "#7fbbb3",
    "bold": true
  },
  "h2": {
    "prefix": "## "
  },
  "h3": {
    "prefix": "### "
  },
  "h4": {
    "prefix": "#### "
  },
  "h5": {
    "prefix": "##### "
  },
  "h6": {
    "prefix": "###### ",
    "color": "#859289",
    "bold": false
  },
  "text": {},
  "strikethrough": {
    "crossed_out": true
  },
  "emph": {
    "italic": true
  },
  "strong": {
    "bold": true
  },
  "hr": {
    "color": "#859289",
    "format": "\n--------\n"
  },
  "item": {
    "block_prefix": "• "
  },
  "enumeration": {
    "block_prefix": ". "
  },
  "task": {
    "ticked": "[✓] ",
    "unticked": "[ ] "
  },
  "link": {
    "color": "#7fbbb3",
    "underline": true
  },
  "link_text": {
    "color": "#83c092",
    "bold": true
  },
  "image": {
    "color": "#7fbbb3",
    "underline": true
  },
  "image_text": {
    "color": "#859289",
    "format": "Image: {{.text}} →"
  },
  "code": {
    "prefix": " ",
    "suffix": " ",
    "color": "#e69875",
    "background_color": "#2e383c"
  },
  "code_block": {
    "color": "#a7c080",
    "margin": 0,
    "chroma": {
      "text": {
        "color": "#d3c6aa"
      },
      "error": {
        "color": "#e67e80"
      },
      "comment": {
        "color": "#859289",
        "italic": true
      },
      "comment_preproc": {
        "color": "#e69875"
      },
      "keyword": {
        "color": "#e67e80"
      },
      "keyword_reserved": {
        "color": "#d699b6"
      },
      "keyword_namespace": {
        "color": "#d699b6"
      },
      "keyword_type": {
        "color": "#dbbc7f"
      },
      "operator": {
        "color": "#7fbbb3"
      },
      "punctuation": {
        "color": "#d3c6aa"
      },
      "name": {
        "color": "#d3c6aa"
      },
      "name_builtin": {
        "color": "#d699b6"
      },
      "name_tag": {
        "color": "#e67e80"
      },
      "name_attribute": {
        "color": "#83c092"
      },
      "name_class": {
        "color": "#dbbc7f",
        "bold": true
      },
      "name_function": {
        "color": "#a7c080"
      },
      "literal_number": {
        "color": "#d699b6"
      },
      "literal_string": {
        "color": "#83c092"
      },
      "literal_string_escape": {
        "color": "#7fbbb3"
      },
      "generic_deleted": {
        "color": "#e67e80"
      },
      "generic_inserted": {
        "color": "#a7c080"
      },
      "generic_emph": {
        "italic": true
      },
      "generic_strong": {
        "bold": true
      },
      "generic_subheading": {
        "color": "#859289"
      },
      "background": {
        "background_color": "#2e383c"
      }
    }
  },
  "table": {},
  "definition_list": {},
  "definition_term": {},
  "definition_description": {},
  "html_block": {},
  "html_span": {}
}`

File: cpp/pl/minisearch/server/markdown.cpp
// Copyright (c) 2026 The Authors. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Authors: liubang (it.liubang@gmail.com)
// Created: 2026/08/17

#include "cpp/pl/minisearch/server/markdown.h"

#include <cctype>
#include <utility>

namespace pl::minisearch::server {

namespace {

std::string trim(const std::string& s) {
    size_t begin = 0;
    while (begin < s.size() && std::isspace(static_cast<unsigned char>(s[begin])) != 0) {
        ++begin;
    }
    size_t end = s.size();
    while (end > begin && std::isspace(static_cast<unsigned char>(s[end - 1])) != 0) {
        --end;
    }
    return s.substr(begin, end - begin);
}

// ATX 标题：^#{1,6}\s+text。返回 (level, text)；非标题返回 (0, "")。
std::pair<int, std::string> parse_heading(const std::string& line) {
    size_t i = 0;
    while (i < line.size() && line[i] == '#') {
        ++i;
    }
    if (i == 0 || i > 6 || i >= line.size() || (line[i] != ' ' && line[i] != '\t')) {
        return {0, ""};
    }
    std::string text = trim(line.substr(i));
    // 去掉结尾的闭合 # 序列（## title ##）
    size_t end = text.size();
    while (end > 0 && text[end - 1] == '#') {
        --end;
    }
    return {static_cast<int>(i), trim(text.substr(0, end))};
}

bool is_fence(const std::string& line) {
    const std::string t = trim(line);
    return t.rfind("```", 0) == 0 || t.rfind("~~~", 0) == 0;
}

// 按空行拆块；fence 代码块作为原子单位（不跨 fence 拆分）。
std::vector<std::string> split_blocks(const std::string& text) {
    std::vector<std::string> blocks;
    std::string current;
    bool in_fence = false;
    size_t pos = 0;
    while (pos <= text.size()) {
        const size_t nl = text.find('\n', pos);
        const std::string line =
            nl == std::string::npos ? text.substr(pos) : text.substr(pos, nl - pos);
        const bool blank = trim(line).empty();
        if (is_fence(line)) {
            in_fence = !in_fence;
        }
        if (blank && !in_fence) {
            if (!trim(current).empty()) {
                blocks.push_back(trim(current));
            }
            current.clear();
        } else {
            if (!current.empty()) {
                current.push_back('\n');
            }
            current += line;
        }
        if (nl == std::string::npos) {
            break;
        }
        pos = nl + 1;
    }
    if (!trim(current).empty()) {
        blocks.push_back(trim(current));
    }
    return blocks;
}

// 单个超长块按 max_chars 硬切（保留 overlap）。
// 起止位置都回退/对齐到 UTF-8 字符边界，避免产出非法 UTF-8 的 chunk。
void hard_split(const std::string& block, const ChunkOptions& opts, std::vector<std::string>* out) {
    const size_t step =
        opts.max_chars > opts.overlap_chars ? opts.max_chars - opts.overlap_chars : opts.max_chars;
    size_t pos = 0;
    while (pos < block.size()) {
        // 起点对齐：跳过上一 chunk 尾部截断残留的 UTF-8 continuation bytes
        while (pos < block.size() && (static_cast<unsigned char>(block[pos]) & 0xC0) == 0x80) {
            ++pos;
        }
        if (pos >= block.size()) {
            break;
        }
        size_t len = opts.max_chars;
        if (pos + len < block.size()) {
            // 终点回退到字符边界
            while (len > 0 && (static_cast<unsigned char>(block[pos + len]) & 0xC0) == 0x80) {
                --len;
            }
        }
        out->push_back(block.substr(pos, len));
        pos += step;
    }
}

// 块列表贪心合并成 <= max_chars 的 chunk；单块超限则硬切。
std::vector<std::string> merge_blocks(const std::vector<std::string>& blocks,
                                      const ChunkOptions& opts) {
    std::vector<std::string> chunks;
    std::string current;
    for (const auto& block : blocks) {
        if (block.size() > opts.max_chars) {
            if (!current.empty()) {
                chunks.push_back(current);
                current.clear();
            }
            hard_split(block, opts, &chunks);
            continue;
        }
        const size_t joined = current.empty() ? block.size() : current.size() + 1 + block.size();
        if (joined > opts.max_chars) {
            chunks.push_back(current);
            current = block;
        } else {
            if (!current.empty()) {
                current.push_back('\n');
            }
            current += block;
        }
    }
    if (!current.empty()) {
        chunks.push_back(std::move(current));
    }
    return chunks;
}

void append_chunks(const std::string& title_path,
                   const std::string& body,
                   const ChunkOptions& opts,
                   std::vector<MarkdownChunk>* out) {
    const std::string trimmed = trim(body);
    if (trimmed.empty()) {
        return;
    }
    for (auto& text : merge_blocks(split_blocks(trimmed), opts)) {
        out->push_back(MarkdownChunk{title_path, std::move(text)});
    }
}

} // namespace

std::vector<MarkdownChunk> ChunkMarkdown(const std::string& markdown, const ChunkOptions& opts) {
    std::vector<MarkdownChunk> chunks;
    // 标题栈：levels[i] 单调递增，path[i] 为对应层级标题文本
    std::vector<int> levels;
    std::vector<std::string> titles;
    std::string body; // 当前 section 的累积正文
    bool in_fence = false;

    auto title_path = [&]() {
        std::string out;
        for (const auto& t : titles) {
            if (!out.empty()) {
                out += " > ";
            }
            out += t;
        }
        return out;
    };

    size_t pos = 0;
    while (pos <= markdown.size()) {
        const size_t nl = markdown.find('\n', pos);
        const std::string line =
            nl == std::string::npos ? markdown.substr(pos) : markdown.substr(pos, nl - pos);
        if (is_fence(line)) {
            in_fence = !in_fence;
        }
        if (!in_fence) {
            const auto [level, text] = parse_heading(line);
            if (level > 0) {
                // 新标题：flush 上一个 section，更新标题栈
                append_chunks(title_path(), body, opts, &chunks);
                body.clear();
                while (!levels.empty() && levels.back() >= level) {
                    levels.pop_back();
                    titles.pop_back();
                }
                levels.push_back(level);
                titles.push_back(text);
                if (nl == std::string::npos) {
                    break;
                }
                pos = nl + 1;
                continue;
            }
        }
        body += line;
        body.push_back('\n');
        if (nl == std::string::npos) {
            break;
        }
        pos = nl + 1;
    }
    append_chunks(title_path(), body, opts, &chunks);
    return chunks;
}

std::vector<MarkdownChunk> ChunkFixed(const std::string& text, const ChunkOptions& opts) {
    std::vector<MarkdownChunk> chunks;
    for (auto& body : merge_blocks(split_blocks(trim(text)), opts)) {
        chunks.push_back(MarkdownChunk{"", std::move(body)});
    }
    return chunks;
}

} // namespace pl::minisearch::server
